### PR TITLE
DHFPROD-5989: Configure workUnit and endpointState in Spark connector

### PR DESCRIPTION
**Pull request by anu3990**
#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [x] JIRA_ID included in all the commit messages
- [x] PR title is in the format JIRA_ID:Title
- [x] Rebase the branch with upstream
- [x] Squashed all commits into a single commit
- [x] Added Tests
  

- ##### Reviewer:

- [x] Reviewed Tests

### CODE REVIEW

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (SameeraPriyathamTadikonda) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:96]: Please remove Sysout

**Comment by SameeraPriyathamTadikonda** [comment]
If you remove the taskID. we should fix the bulkIngesterTest. It is better if we have two different PR's/Commits.
1) For configuring customEndpointState and Workunits
2) Removing taskId and moving the URIPrefix to WorkUnit

**Review by SameeraPriyathamTadikonda — CHANGES_REQUESTED** [review]

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:70]: There's a lot going on here, and generally, I like to use private/protected methods to break up long methods so that they're easier to read and maintain. I prefer private, though I'll use protected when I want to write unit tests against the logic.

I see the following steps in this code, where each can become a new private method:

1. Use the user params to determine the apiPath, the workUnit, and endpointState (and throw a validation error as needed)
2. Use the output of #1 to create a BulkInputCaller

I think the signature of the first private method would be "private JsonNode determineIngestionEndpointParams(Map)". If the user has defined "ingestendpointparams", then use ObjectMapper to parse that into a JsonNode, and then validate it. Else, use objectMapper.createObjectNode() and populate apiPath based on the default API path, and then populate workUnit based on uriPrefix (and then @SameeraPriyathamTadikonda  would add more logic to this for collections/permissions/etc). 

That first method would likely be protected so that you could easily write unit tests against it without having to involve everything else in this class. 

The second private method would have a signature of "private BulkInputCaller buildBulkInputCaller(JsonNode ingestionParams)". 

Each private method can then have a try/catch as needed to ensure that if an error is thrown, the appropriate context is provided in the error message.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriterFactory.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriterFactory.java:56]: I think that since taskId doesn't matter, let's remove the debug logging above it. It would be meaningless to see it.

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java:99]: When there are a lot of params in a method, that's generally a sign that a class should be created. 

While we don't need this class in the application code, an "Options" class would be very helpful here for making it easier to write tests. It would have properties of batchSize, uriPrefix, apiPath, etc. And then it would have a "Map toMap()" method that builds a Map of params based on what's been configured. 

Then, each test will just build an Options class based on the scenario that it's testing. That will make it easier to write future tests - i.e. when @SameeraPriyathamTadikonda  needs to test for collections/permissions/etc, we don't have to add those to this method and fix every method that calls it - we just add new properties to our Options class, and the existing tests won't be affected.

**Review by rjrudin — CHANGES_REQUESTED** [review]

**Comment by anu3990** [comment]
@SameeraPriyathamTadikonda . for your below comments - 
If you remove the taskID. we should fix the bulkIngesterTest. It is better if we have two different PR's/Commits.
For configuring customEndpointState and Workunits
Removing taskId and moving the URIPrefix to WorkUnit

1. i have fixed bulkIngesterTest but the user can still have a field named "taskId" in their workUnit. I have renamed it for now to avoid confusion.
2. For your second point - "Removing taskId and moving the URIPrefix to WorkUnit", this is exactly what we do when we say "configure workUnit" so we can still combine it in the current story - https://project.marklogic.com/jira/browse/DHFPROD-5989 because we are configuring both endpointState and workUnit here. I have also added a bullet point in the acceptance criteria of the story.

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java:37]: Pinging @SameeraPriyathamTadikonda  about this - for each workUnit input that the endpoint supports, we'll want to verify that it works correctly in the BulkIngestTest class, as that class has no knowledge of Spark.

The tests for our Spark connector should then focus on converting the params Map into the proper workUnit object. Those tests don't need to connect to ML - they can be very fast unit tests that live in a separate class that doesn't extend AbstractHubCoreTest. I'll put together a PR to demonstrate this and submit it against this branch. 

We'll still want one test in this project that writes data with every workUnit field populated, just to make sure that we're setting the correct workUnit fields.

File: marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java
Comment (rjrudin) [inline marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java:35]: Can this just be "{}", since we don't care about "next"? It's supposed to go away once we shift to the newer Java client, but I think it'd be better to make it "{}" in the meantime.

**Review by rjrudin — COMMENTED** [review]
I'm going to submit a separate PR with a test helper class.

**Comment by rjrudin** [comment]
@anu3990  @SameeraPriyathamTadikonda See this PR - https://github.com/anu3990/marklogic-data-hub/pull/2 - for an Options class that should simplify writing tests. It also facilitates testing for null inputs, as  options in the map will only be populated if the test populates them.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:174]: I am wondering if the ResourceNotFoundException needs to be caught here. If we do throw a new exception, we definitely need to include the original exception so that Ernie doesn't lose any context. 

I did a quick test, and this is what the exception shows:

```
com.marklogic.client.ResourceNotFoundException: Local message: Could not read non-existent document. Server Message: RESTAPI-NODOCUMENT: (err:FOER0000) Resource or document does not exist:  category: content message: /data-hub/5/data-services/ingestion/bulkIngesterrrr.api
```

That is likely helpful enough for Ernie - it shows the module he referenced, and the Java Client is giving a good indication of the error. So I think this try/catch can be removed. 

And because of that, we don't really need this method anymore, since it's just a few lines of code now. If we do need some special error handling here, I think it's worth the separate method. But that doesn't appear to be the case.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:132]: Exception messages are crucial here to explain to Ernie what to do if he made a mistake. We want to be more precise about how to fix the problem - e.g. 

Cannot set workUnit or endpointState in ingestionendpointparams unless apiPath is defined as well

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:59]: No need to capture this here - batchSize needs to be captured, but not this. Just need to read this in the determineIngestionEndpointParams method.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:158]: Use set("workUnit", workUnitNode) instead. That returns the JSON structure instead of converting it into a string. You can then do setWorkUnit(new JacksonHandle(theNode)) and don't have to fiddle with converting things into strings. Same goes for endpointState.

**Review by rjrudin — CHANGES_REQUESTED** [review]

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:163]: I think this needs to be a JsonNode as well instead of a string for consistency. You can do mapper.createObjectNode() and then node.put("uriprefix", params.get("uriprefix")) . 

I think it's good to handle these as JsonNode's instead of as strings. When Ernie provides us with a custom workUnit and/or endpointState, we should convert those ASAP to JsonNode. That way, if Ernie's JSON is malformed, it'll fail as quickly as possible - i.e. in the connector code and not in ML. 

Also this is where @SameeraPriyathamTadikonda will add his code for collections/permissions/etc.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:66]: I'll test this locally, but I didn't think "asText" does what we want here - it's just getting the text of this node, it's not the same as toString. I think we want to use JacksonHandle here instead, but I'm not sure. 

Which makes me realize - we really need a test that uses a custom API path. That way, we know for sure that our custom workUnit and endpointState are passed to our custom endpoint correctly.

Let me handle that test - I want to think a bit about where the modules for this should go. I think they can go under ./marklogic-data-hub/src/test/ml-modules, I just need to confirm that. I'll try to submit that tonight or first thing tomorrow morning. 

In the meantime though, this can go forward so that @SameeraPriyathamTadikonda  can make use of it.

**Review by rjrudin — APPROVED** [review]

File: marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java
Comment (SameeraPriyathamTadikonda) [inline marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java:36]: what's the purpose of userDefinedValue?

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (SameeraPriyathamTadikonda) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:163]: Do we need to add this even if its null?

File: marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java
Comment (anu3990) [reply marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java:36]: "userDefinedValue" is just a variable added by the user. We can also have something like the below -
String workUnit      = "{\"testValue\":" + 1 + ", \"uriprefix\":\""+prefix+"\"}";

Adding the "userDefinedValue" confirms that no exception is thrown even when we have extra values defined by the user.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (anu3990) [reply marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:163]: Yes we need this even when the user did not add URIPrefix because our sjs file looks for this value. Since, the uriPrefix value does not change, we moved it out from endpointState to workUnit

Comment (SameeraPriyathamTadikonda) [reply marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:163]: Are we going to add null in that case?

File: marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java
Comment (SameeraPriyathamTadikonda) [reply marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java:36]: How will the user pass this option from glue/spark? I'm still confused where are we using this?

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (anu3990) [reply marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:163]: No we will not be adding null to the uris. Our Glue script will be something like this -

val prefix = ""
val writer = dataframe.write.format("com.marklogic.client.spark.Writer.MarkLogicWriteDataSource")
                    .option("host", host)
                    .option("port", port)
                    .option("user", username)
                    .option("password",password)
                    .option("uriprefix", prefix).....(other options);
By default the value of variable prefix is an empty string. The user can add a custom value.

File: marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java
Comment (anu3990) [reply marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java:36]: This test is not in the marklogic-data-hub-spark-connector folder. It is testing only the bulk caller in the java-client-api and uses the bulkIngester api and sjs. The uriPrefix appended in the workUnit is used by the sjs file here - https://github.com/anu3990/marklogic-data-hub/blob/660d87d7616098274a5b97ff3d2005b0a07461fb/marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/ingestion/bulkIngester.sjs#L39 

In the Glue environment, uriprefix will be passed in as an option(i have defined in the next comment) and appended in the workUnit by the HubDataWriter class before calling bulkIngester.api.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (rjrudin) [reply marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:163]: But I don't think Ernie has to specify '.option("uriprefix", prefix)', right? In which case the option will not exist in the options map? And so we'd only want to add it to the workUnit if it's not null (unless the endpoint is careful to not use it if it's null).

Comment (SameeraPriyathamTadikonda) [reply marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:163]: Yes, that's exactly my point. I don't think that is handled in the endpoint.
Also the endpoint throws an exception if workUnit is empty when no options are set by the user.

File: marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java
Comment (SameeraPriyathamTadikonda) [reply marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java:36]: I understand that. I'm confused on the userDefinedValue. How it is passed and what's the use of it?

**Comment by rjrudin** [comment]
@anu3990  @SameeraPriyathamTadikonda  I created https://github.com/anu3990/marklogic-data-hub/pull/3 to add a test that uses a custom endpoint. I moved some methods from WriteDataTest to the parent class so that it's easier to write separate test classes, as we don't want every single scenario going into WriteDataTest.

I made a couple fixes to HubDataWriter so that workUnit and endpointState are always ObjectNodes, which means we can use JacksonHandle when passing them to the BulkInputCaller. All tests are passing.

**Review by rjrudin — CHANGES_REQUESTED** [review]
Will review again after above PR is merged in.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (SameeraPriyathamTadikonda) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:138]: endpointParams.get("apiPath").asText() will throw s NPE

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (SameeraPriyathamTadikonda) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:144]: we can use the variable "doesNotHaveApiPath" instead of writing the logic again

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (SameeraPriyathamTadikonda) [inline marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:148]: else block missing

File: marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java
Comment (SameeraPriyathamTadikonda) [reply marklogic-data-hub/src/test/java/com/marklogic/hub/dataservices/ingestion/BulkIngestTest.java:36]: We don't need the UserDefinedValue here.

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataViaCustomEndpointTest.java
Comment (SameeraPriyathamTadikonda) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataViaCustomEndpointTest.java:22]: provide a meaningful name for this test

**Comment by SameeraPriyathamTadikonda** [comment]
Will the ingestionEndpointParams be sent as a JSON String from glue to the connector?

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (rjrudin) [reply marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:138]: An NPE won't occur because if apiPath were null or did not exist, then the first part of the conditional would be true and then Java won't evaluate the second part.

File: marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java
Comment (rjrudin) [reply marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java:148]: There's no else needed - this is just saying "If Ernie didn't provide a non-null endpointState, then toss in an empty object node".

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataViaCustomEndpointTest.java
Comment (rjrudin) [reply marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataViaCustomEndpointTest.java:22]: This is actually a common pattern we have in existing tests - since the class name already identifies the scope of the test, the method is just "test". If a second test method is added, then this would need to be updated.

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java:78]: I think you can drop this test because WriteDataViaCustomEndpointTest already covers this with an actual custom endpoint, as opposed to using the default endpoint.

**Review by rjrudin — APPROVED** [review]
Looks good to me, though I think it's worth dropping ingestWithCustomApiWithCustomWorkunit in favor of the separate test class that tests a custom endpoint.

**Review by SameeraPriyathamTadikonda — DISMISSED** [review]

**Comment by anu3990** [comment]
@rjrudin and @SameeraPriyathamTadikonda. Need your approvals again, i have rebased and added the suggestions.

**Review by rjrudin — APPROVED** [review]

**Review by SameeraPriyathamTadikonda — APPROVED** [review]

**Pull request merged**
